mithril-signer/src/mktree_store_sqlite.rs

```diff
@@ -0,0 +1,169 @@
+use std::{
+    iter::repeat,
+    sync::{Arc, RwLock},
+};
+
+use anyhow::Context;
+use mithril_common::{
+    crypto_helper::{
+        ckb_merkle_mountain_range::{
+            Error as MMRError, MMRStoreReadOps, MMRStoreWriteOps, Result as MMRResult,
+        },
+        Bytes, MKTreeLeafIndexer, MKTreeLeafPosition, MKTreeNode, MKTreeStorer,
+    },
+    StdResult,
+};
+
+/// A Merkle tree store with Sqlite backend
+/// This store is used to store the Merkle tree nodes in a Sqlite database with in memory mode.
+/// This store is not suited for proving, it is only used to store the compute the root of a Merkle tree.
+/// This store is slower than the in memory store but uses less memory which is important to minimize the signer footprint on the SPO infrastructure.
+pub struct MKTreeStoreSqlite {
+    inner_store: RwLock<sqlite::ConnectionThreadSafe>,
+}
+
+impl MKTreeStoreSqlite {
+    fn build() -> StdResult<Self> {
+        Ok(Self {
+            inner_store: RwLock::new(Self::create_connection()?),
+        })
+    }
+
+    fn create_connection() -> StdResult<sqlite::ConnectionThreadSafe> {
+        let connection = sqlite::Connection::open_thread_safe(":memory:")?;
+        connection.execute("pragma shrink_memory; pragma temp_store = FILE;")?;
+        connection.execute(
+            "create table merkle_tree (
+                position integer, 
+                element blob, 
+                primary key (position)
+            )",
+        )?;
+
+        Ok(connection)
+    }
+
+    fn get_element_at_position(&self, position: u64) -> StdResult<Option<Arc<MKTreeNode>>> {
+        let inner_store = self.inner_store.read().unwrap();
+        let query = "SELECT element FROM merkle_tree WHERE position = ?";
+        let mut statement = (*inner_store).prepare(query)?;
+        statement.bind((1, position as i64)).unwrap();
+        let result = if let Ok(sqlite::State::Row) = statement.next() {
+            Some(Arc::new(MKTreeNode::new(
+                statement.read::<Bytes, _>("element")?,
+            )))
+        } else {
+            None
+        };
+
+        Ok(result)
+    }
+
+    fn insert_elements_from_position(
+        &self,
+        position: u64,
+        elements: Vec<Arc<MKTreeNode>>,
+    ) -> StdResult<()> {
+        let inner_store = self.inner_store.read().unwrap();
+        let values_columns: Vec<&str> = repeat("(?, ?)").take(elements.len()).collect();
+        let values: Vec<sqlite::Value> =
+            elements
+                .into_iter()
+                .enumerate()
+                .fold(vec![], |mut vec, (i, elem)| {
+                    vec.append(&mut vec![
+                        sqlite::Value::Integer((position + i as u64) as i64),
+                        sqlite::Value::Binary((**elem).to_vec()),
+                    ]);
+                    vec
+                });
```

Comment:
I would prefer to use `flat_map` here as it's a simpler operation than `fold`.
```suggestion
                .flat_map(|(i, elem)| {
                vec![
                    sqlite::Value::Integer((position + i as u64) as i64),
                    sqlite::Value::Binary((**elem).to_vec()),
                ]
            })
            .collect();
```

Reply:
Done in https://github.com/input-output-hk/mithril/pull/1904/commits/8da2bfe8d276e5028c2f0a718c0ab1e6c4dc9e24